### PR TITLE
GNNPolicyApproach

### PR DESCRIPTION
**Pull request by ronuchit**
Performs badly with our regular environment parameters, but does
decently on some simplified settings:

*Cover*
Settings changes:
```
    cover_num_blocks = 1
    cover_num_targets = 1
    cover_block_widths = [0.1]
    cover_target_widths = [0.05]
```
Command: `python src/main.py --env cover --approach gnn_policy --seed 0`
Results: 38/50

*Blocks*
Settings changes:
```
    num_blocks_train = [3]
    num_blocks_test = [3]
```
Command: `python src/main.py --env blocks --approach gnn_policy --seed
0`
Results: 47/50

*Painting*
Settings changes:
```
    painting_num_objs_train = [1]
    painting_num_objs_test = [1]
```
Command: `python src/main.py --env painting --approach gnn_policy --seed
0`
Results: 10/50

*Tools*
Settings changes:
```
    tools_num_items_train = [1]
    tools_num_items_test = [1]
    tools_num_contraptions_train = [1]
    tools_num_contraptions_test = [1]
```
Command: `python src/main.py --env tools --approach gnn_policy --seed 0`
Results: 19/50

### CODE REVIEW

File: src/settings.py [hunk 1]
```diff
@@ -87,9 +87,17 @@ class GlobalSettings:
...
     behavior_randomize_init_state = False
 
-    # parameters for approaches
+    # parameters for random options approach
     random_options_max_tries = 100
 
+    # parameters for GNN policy approach
+    gnn_policy_num_message_passing = 3
+    gnn_policy_learning_rate = 1e-3
+    gnn_policy_num_epochs = 2500
```
Comment (ronuchit) [inline src/settings.py:96]: @tomsilver this parameter is very important. it used to be 500 in other codebases, but increasing it to 2500 helped a LOT

File: src/approaches/gnn_policy_approach.py [hunk 2]
```diff
@@ -0,0 +1,497 @@
...
+        self._input_normalizers = info["input_normalizers"]
+        self._target_normalizers = info["target_normalizers"]
+
+    def _predict_option(self, state: State, atoms: Set[GroundAtom],
+                        goal: Set[GroundAtom]) -> _Option:
+        # Get output graph.
+        in_graph, object_to_node = self._graphify_single_input(
+            state, atoms, goal)
+        node_to_object = {v: k for k, v in object_to_node.items()}
+        type_to_obj = defaultdict(set)
```
Comment (tomsilver) [inline src/approaches/gnn_policy_approach.py:209]: rename to `type_to_node`?

File: src/approaches/gnn_policy_approach.py [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+        # Extract objects, making sure to enforce the typing constraints.
+        objects = []
+        for i, obj_type in enumerate(param_opt.types):
+            scores = out_graph["nodes"][:, i]
+            allowed_idxs = type_to_obj[obj_type.name]
+            for j in range(len(scores)):
+                if j not in allowed_idxs:
+                    scores[j] = float("-inf")  # set its score to be really bad
+            if np.max(scores) == float("-inf"):  # type: ignore
+                # If all scores are None, we failed to select an object.
```
Comment (tomsilver) [inline src/approaches/gnn_policy_approach.py:239]: is this actually possible? also instead of None, you mean -inf, right?

File: src/approaches/gnn_policy_approach.py [hunk 4]
```diff
@@ -0,0 +1,497 @@
...
+
+        # Go through the data, identifying the maximum number of option
+        # objects and parameters, and the types/predicates/attributes.
+        for state, atoms, goal, option in data:
+            assert len(option.params.shape) == 1
+            max_option_objects = max(max_option_objects, len(option.objects))
+            max_option_params = max(max_option_params, option.params.shape[0])
+            for atom in atoms | goal:
+                arity = atom.predicate.arity
+                assert arity <= 2
```
Comment (tomsilver) [inline src/approaches/gnn_policy_approach.py:264]: leave an informative error message?

File: src/settings.py (hunk 1 above)
Comment (tomsilver) [reply src/settings.py:96]: good to know

File: src/approaches/gnn_policy_approach.py [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+            option_index = self._sorted_options.index(option.parent)
+            onehot_target = np.zeros(len(self._sorted_options))
+            onehot_target[option_index] = 1
+            assert len(option.params.shape) == 1
+            params_target = np.zeros(self._max_option_params)
+            params_target[:option.params.shape[0]] = option.params
+            graph_target["globals"] = np.r_[onehot_target, params_target]
+            graph_targets.append(graph_target)
+
+        if CFG.gnn_policy_do_normalization:
```
Comment (tomsilver) [inline src/approaches/gnn_policy_approach.py:357]: this will happen twice, first for the exemplar creation, and then for the full dataset. the second will overwrite the first, which is good, because we want to normalize with the full dataset. but maybe should clarify/fix this anyway.

File: requirements.txt [hunk 6]
```diff
@@ -13,3 +13,4 @@ pyperplan
 pathos
 requests
 slack_bolt
+networkx
```
[the remark this answers is not in the document]
Comment (ronuchit) [reply requirements.txt:16]: i'll remove the dependency

[the remark this answers is not in the document]
Comment (ronuchit) [reply requirements.txt:16]: it's in `gnn/` for visualization stuff

File: src/approaches/gnn_policy_approach.py [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+            example_dataset,
+            use_gpu=False,
+            num_steps=CFG.gnn_policy_num_message_passing)
+        # Set up all the graphs, now using *all* the data.
+        inputs, targets = self._graphify_data([(d[0], d[1], d[2])
+                                               for d in data],
+                                              [d[3] for d in data])
+        # Run training.
+        if CFG.gnn_policy_use_validation_set:
+            ## Split data, using 10% for validation.
```
[the remark this answers is not in the document]
Comment (ronuchit) [reply src/approaches/gnn_policy_approach.py:104]: an "indented" comment that's under a parent

File: src/approaches/gnn_policy_approach.py [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+            "nullary_predicates": self._nullary_predicates,
+            "max_option_objects": self._max_option_objects,
+            "max_option_params": self._max_option_params,
+            "node_feature_to_index": self._node_feature_to_index,
+            "edge_feature_to_index": self._edge_feature_to_index,
+            "input_normalizers": self._input_normalizers,
+            "target_normalizers": self._target_normalizers,
+        }
+        save_path = utils.get_approach_save_path_str()
+        with open(f"{save_path}_None.gnn", "wb") as f:
```
[the remark this answers is not in the document]
Comment (ronuchit) [reply src/approaches/gnn_policy_approach.py:179]: we've always had that, e.g. `tools__nsrt_learning__459__all__tools_noinventallexclude_200demo.saved_None.NSRTs`

File: src/approaches/gnn_policy_approach.py (hunk 2 above)
Comment (ronuchit) [reply src/approaches/gnn_policy_approach.py:209]: good catch!

File: src/approaches/gnn_policy_approach.py (hunk 3 above)
Comment (ronuchit) [reply src/approaches/gnn_policy_approach.py:239]: yeah it was happening in a couple episodes in tools, i didn't investigate much bc it was rare enough

**Comment by ronuchit** [comment]
Results on 2500 epochs:
```
Git commit hashes seen in results/:
9fa857c52027d30ef460420e9ce79ee3f837c261


AGGREGATED DATA OVER SEEDS:
                           NUM_TRAIN_TASKS    NUM_SOLVED AVG_NUM_PREDS AVG_TEST_TIME AVG_NODES_CREATED    LEARNING_TIME  NUM_SEEDS
EXPERIMENT_ID                                                                                                                     
blocks_gnn_exclude_100       100.00 (0.00)  25.80 (5.19)   0.00 (0.00)   0.04 (0.01)       0.00 (0.00)   233.42 (10.26)         10
blocks_gnn_exclude_150       150.00 (0.00)  27.00 (6.28)   0.00 (0.00)   0.04 (0.01)       0.00 (0.00)   346.07 (27.22)         10
blocks_gnn_exclude_200       200.00 (0.00)  29.10 (5.54)   0.00 (0.00)   0.05 (0.01)       0.00 (0.00)   439.88 (17.05)         10
blocks_gnn_exclude_50         50.00 (0.00)  14.10 (6.17)   0.00 (0.00)   0.04 (0.01)       0.00 (0.00)   140.44 (13.36)         10
blocks_gnn_noexclude_100     100.00 (0.00)  18.50 (5.37)   0.00 (0.00)   0.05 (0.01)       0.00 (0.00)   245.80 (20.91)         10
blocks_gnn_noexclude_150     150.00 (0.00)  21.80 (5.23)   0.00 (0.00)   0.05 (0.01)       0.00 (0.00)   351.99 (27.92)         10
blocks_gnn_noexclude_200     200.00 (0.00)  27.20 (4.94)   0.00 (0.00)   0.05 (0.01)       0.00 (0.00)   444.80 (40.92)         10
blocks_gnn_noexclude_50       50.00 (0.00)  15.50 (4.50)   0.00 (0.00)   0.05 (0.01)       0.00 (0.00)    135.76 (9.29)         10
cover_gnn_exclude_100        100.00 (0.00)   1.50 (0.67)   0.00 (0.00)   0.00 (0.00)       0.00 (0.00)    103.07 (6.79)         10
cover_gnn_exclude_150        150.00 (0.00)   1.60 (1.50)   0.00 (0.00)   0.00 (0.00)       0.00 (0.00)   125.77 (10.40)         10
cover_gnn_exclude_200        200.00 (0.00)   1.10 (1.30)   0.00 (0.00)   0.00 (0.00)       0.00 (0.00)   168.26 (16.06)         10
cover_gnn_exclude_50          50.00 (0.00)   0.50 (0.67)   0.00 (0.00)   0.00 (0.00)       0.00 (0.00)     62.52 (3.22)         10
cover_gnn_noexclude_100      100.00 (0.00)   4.00 (2.19)   0.00 (0.00)   0.02 (0.03)       0.00 (0.00)    124.13 (8.89)         10
cover_gnn_noexclude_150      150.00 (0.00)   4.00 (2.14)   0.00 (0.00)   0.01 (0.00)       0.00 (0.00)   176.78 (11.33)         10
cover_gnn_noexclude_200      200.00 (0.00)   5.90 (3.83)   0.00 (0.00)   0.01 (0.00)       0.00 (0.00)   231.42 (19.41)         10
cover_gnn_noexclude_50        50.00 (0.00)   1.90 (1.30)   0.00 (0.00)   0.00 (0.00)       0.00 (0.00)     75.82 (7.06)         10
painting_gnn_exclude_100     100.00 (0.00)   0.00 (0.00)   0.00 (0.00)     nan (nan)       0.00 (0.00)   515.54 (34.11)         10
painting_gnn_exclude_150     150.00 (0.00)   0.00 (0.00)   0.00 (0.00)     nan (nan)       0.00 (0.00)   743.63 (39.66)         10
painting_gnn_exclude_200     200.00 (0.00)   0.00 (0.00)   0.00 (0.00)     nan (nan)       0.00 (0.00)  1003.30 (51.85)         10
painting_gnn_exclude_50       50.00 (0.00)   0.00 (0.00)   0.00 (0.00)     nan (nan)       0.00 (0.00)    263.73 (9.41)         10
painting_gnn_noexclude_100   100.00 (0.00)   0.10 (0.30)   0.00 (0.00)   0.08 (0.00)       0.00 (0.00)   527.36 (35.10)         10
painting_gnn_noexclude_150   150.00 (0.00)   0.00 (0.00)   0.00 (0.00)     nan (nan)       0.00 (0.00)   776.59 (45.96)         10
painting_gnn_noexclude_200   200.00 (0.00)   0.00 (0.00)   0.00 (0.00)     nan (nan)       0.00 (0.00)  1010.37 (42.11)         10
painting_gnn_noexclude_50     50.00 (0.00)   0.10 (0.30)   0.00 (0.00)   0.12 (0.00)       0.00 (0.00)   278.64 (20.53)         10
tools_gnn_exclude_100        100.00 (0.00)   1.20 (0.87)   0.00 (0.00)   0.04 (0.01)       0.00 (0.00)   427.45 (28.84)         10
tools_gnn_exclude_150        150.00 (0.00)   2.90 (2.12)   0.00 (0.00)   0.04 (0.00)       0.00 (0.00)   629.55 (46.79)         10
tools_gnn_exclude_200        200.00 (0.00)   4.80 (2.27)   0.00 (0.00)   0.04 (0.00)       0.00 (0.00)   783.05 (33.30)         10
tools_gnn_exclude_50          50.00 (0.00)   0.00 (0.00)   0.00 (0.00)     nan (nan)       0.00 (0.00)   221.61 (12.58)         10
tools_gnn_noexclude_100      100.00 (0.00)   3.00 (3.10)   0.00 (0.00)   0.04 (0.01)       0.00 (0.00)    444.56 (5.32)         10
tools_gnn_noexclude_150      150.00 (0.00)   5.20 (2.82)   0.00 (0.00)   0.07 (0.08)       0.00 (0.00)   676.95 (24.82)         10
tools_gnn_noexclude_200      200.00 (0.00)   7.00 (2.68)   0.00 (0.00)   0.04 (0.00)       0.00 (0.00)   871.17 (27.53)         10
tools_gnn_noexclude_50        50.00 (0.00)   0.10 (0.30)   0.00 (0.00)   0.04 (0.00)       0.00 (0.00)    235.62 (7.80)         10
```

**Comment by ronuchit** [comment]
Results on 25000 epochs:
```
Git commit hashes seen in results/:
d899e8c57584b88e616d742d95ff5a49cebf1749


AGGREGATED DATA OVER SEEDS:
                         NUM_TRAIN_TASKS    NUM_SOLVED AVG_NUM_PREDS AVG_TEST_TIME AVG_NODES_CREATED     LEARNING_TIME  NUM_SEEDS
EXPERIMENT_ID
blocks_gnn_exclude_100     100.00 (0.00)  25.20 (8.58)   0.00 (0.00)   0.07 (0.10)       0.00 (0.00)  2270.21 (357.34)         10
blocks_gnn_exclude_150     150.00 (0.00)  24.00 (5.39)   0.00 (0.00)   0.05 (0.02)       0.00 (0.00)  3244.03 (228.90)         10
blocks_gnn_exclude_200     200.00 (0.00)  28.80 (2.71)   0.00 (0.00)   0.20 (0.13)       0.00 (0.00)  4261.34 (278.39)         10
blocks_gnn_exclude_50       50.00 (0.00)  23.30 (7.58)   0.00 (0.00)   0.05 (0.03)       0.00 (0.00)   1128.30 (56.82)         10
cover_gnn_exclude_100      100.00 (0.00)   2.20 (2.18)   0.00 (0.00)   0.01 (0.01)       0.00 (0.00)    832.35 (46.88)         10
cover_gnn_exclude_150      150.00 (0.00)   3.00 (1.95)   0.00 (0.00)   0.00 (0.00)       0.00 (0.00)   1028.39 (37.58)         10
cover_gnn_exclude_200      200.00 (0.00)   3.50 (2.01)   0.00 (0.00)   0.00 (0.00)       0.00 (0.00)  1602.58 (346.35)         10
cover_gnn_exclude_50        50.00 (0.00)   0.50 (0.50)   0.00 (0.00)   0.00 (0.00)       0.00 (0.00)    469.94 (35.73)         10
painting_gnn_exclude_100   100.00 (0.00)   0.60 (0.80)   0.00 (0.00)   0.07 (0.01)       0.00 (0.00)  4759.04 (110.68)         10
painting_gnn_exclude_150   150.00 (0.00)   0.60 (0.66)   0.00 (0.00)   0.05 (0.01)       0.00 (0.00)  7397.21 (475.95)         10
painting_gnn_exclude_200   200.00 (0.00)   0.30 (0.46)   0.00 (0.00)   0.07 (0.00)       0.00 (0.00)  9597.33 (619.52)         10
painting_gnn_exclude_50     50.00 (0.00)   0.50 (0.67)   0.00 (0.00)   0.06 (0.01)       0.00 (0.00)  2437.26 (190.92)         10
tools_gnn_exclude_100      100.00 (0.00)   3.00 (1.48)   0.00 (0.00)   0.62 (1.19)       0.00 (0.00)  3894.67 (100.26)         10
tools_gnn_exclude_150      150.00 (0.00)   7.20 (3.16)   0.00 (0.00)   0.21 (0.27)       0.00 (0.00)  5974.27 (264.14)         10
tools_gnn_exclude_200      200.00 (0.00)  10.90 (3.78)   0.00 (0.00)   0.35 (0.36)       0.00 (0.00)  7847.80 (251.61)         10
tools_gnn_exclude_50        50.00 (0.00)   0.50 (1.02)   0.00 (0.00)   0.04 (0.01)       0.00 (0.00)  2122.77 (478.63)         10
```
